dali/pipeline/operators/op_schema.cc

```diff
@@ -101,4 +100,65 @@ std::map<std::string, std::pair<std::string, Value*>> OpSchema::GetOptionalArgum
   return ret;
 }
 
+std::string OpSchema::GetArgumentDox(const std::string &name) const {
+  DALI_ENFORCE(HasArgument(name), "Argument \"" + name + "\" is not supported by operator \"" + this->name() + "\".");
+  if (HasRequiredArgument(name)) {
+    return GetRequiredArguments().at(name).first;
+  } else {
+    // optional argument
+    return GetOptionalArguments().at(name).first;
+  }
+}
+
+DALIDataType OpSchema::GetArgumentType(const std::string &name) const {
+  DALI_ENFORCE(HasArgument(name), "Argument \"" + name + "\" is not supported by operator \"" + this->name() + "\".");
+  if (HasRequiredArgument(name)) {
+    return GetRequiredArguments().at(name).second;
+  } else {
+    // optional argument
+    return GetOptionalArguments().at(name).second->GetTypeID();
+  }
+}
+
+std::string OpSchema::GetArgumentDefaultValueString(const std::string &name) const {
+  DALI_ENFORCE(HasOptionalArgument(name), "Argument \"" + name +
+      "\" is either not supported by operator \"" + this->name() + "\" or is not optional.");
+  return GetOptionalArguments().at(name).second->ToString();
+}
+
+std::vector<std::string> OpSchema::GetArgumentNames() const {
+  std::vector<std::string> ret;
+  for (auto &arg_pair : GetRequiredArguments()) {
+    ret.push_back(arg_pair.first);
+  }
+  for (auto &arg_pair : GetOptionalArguments()) {
+    ret.push_back(arg_pair.first);
+  }
+  return ret;
+}
+
+bool OpSchema::HasRequiredArgument(const std::string &name) const {
+  bool ret = arguments_.find(name) != arguments_.end();
+  if (ret) {
+    return ret;
+  }
+  for (const auto &p : parents_) {
```

Comment:
Maybe guard against cyclic dependency?

Reply:
That will be hard without actually creating either the list of schemas already seen or list of arguments already seen (so that you could check for duplicates). 

Adding test that would catch that (and making docs would) is enough I think.

Reply:
OK. Sounds reasonable. How about adding some note that addParent doesn't support cyclic dependency.

Reply:
Sure, I will add that.